[RenderStyleGen] Make BorderImage/MaskBorder storage more consistent with other RenderStyle types

weinig opened the pull request:
#### 3162815bf526a6e22a97d24d23c0fc2374e0ec2d
<pre>
[RenderStyleGen] Make BorderImage/MaskBorder storage more consistent with other RenderStyle types
<a href="https://bugs.webkit.org/show_bug.cgi?id=304128">https://bugs.webkit.org/show_bug.cgi?id=304128</a>

Reviewed by Darin Adler.

Style::BorderImage and Style::MaskImage had a non-traditional
storage implementation by using a private DataRef value. This
required using `friend` to participate in autogeneration, but
that is not going to scale with additional generation. Instead,
we extract the internal Data types into new `StyleBorderImageData`
and `StyleMaskImageData` types, and invert the relationship.

This also allows us to get rid of the custom style building
code for border image and mask image, and will make it easier
for future automation efforts.

While doing this, I also made the members of the border types public
(and remove the m_ prefixes). We want to treat RenderStyle data
types as pure storage going forward, so being public makes sense.

Did a little cleanup RenderStyle reset functions to remove unused
ones and make the rest call to a set function.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/layout/formattingContexts/block/BlockMarginCollapse.cpp:
* Source/WebCore/page/LocalFrameView.cpp:
* Source/WebCore/rendering/OutlinePainter.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderTable.cpp:
* Source/WebCore/rendering/RenderTableSection.cpp:
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderTreeAsText.cpp:
* Source/WebCore/rendering/style/BorderData.cpp:
* Source/WebCore/rendering/style/BorderData.h:
* Source/WebCore/rendering/style/BorderValue.cpp:
* Source/WebCore/rendering/style/BorderValue.h:
* Source/WebCore/rendering/style/CollapsedBorderValue.h:
* Source/WebCore/rendering/style/OutlineValue.cpp:
* Source/WebCore/rendering/style/OutlineValue.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleBase.h:
* Source/WebCore/rendering/style/RenderStyleBaseInlines.h:
* Source/WebCore/rendering/style/RenderStyleBaseSetters.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStylePropertiesGettersCustom.cpp:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleBackgroundData.cpp:
* Source/WebCore/rendering/style/StyleBorderImageData.cpp: Added.
* Source/WebCore/rendering/style/StyleBorderImageData.h: Copied from Source/WebCore/rendering/style/RenderStylePropertiesGettersCustom.cpp.
* Source/WebCore/rendering/style/StyleMaskBorderData.cpp: Added.
* Source/WebCore/rendering/style/StyleMaskBorderData.h: Copied from Source/WebCore/rendering/style/RenderStylePropertiesGettersCustom.cpp.
* Source/WebCore/rendering/style/StyleMultiColData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleSurroundData.cpp:
* Source/WebCore/style/StyleBuilder.cpp:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleChangedAnimatableProperties.cpp:
* Source/WebCore/style/StyleDifference.cpp:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/values/backgrounds/StyleBorderImage.cpp:
* Source/WebCore/style/values/backgrounds/StyleBorderImage.h:
* Source/WebCore/style/values/backgrounds/StyleBorderImageWidth.h:
* Source/WebCore/style/values/masking/StyleMaskBorder.cpp:
* Source/WebCore/style/values/masking/StyleMaskBorder.h:
* Source/WebCore/style/values/non-standard/StyleWebKitBoxReflect.cpp:

Canonical link: <a href="https://commits.webkit.org/304482@main">https://commits.webkit.org/304482@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92b26673617c41abb4555bbd38e8f863603f887f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135673 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8047 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46965 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143389 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/87331 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137542 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8691 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7892 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103684 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/71649 "Build is in progress. Recent messages:") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138619 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6266 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121621 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84557 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/6032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3647 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3996 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115257 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39807 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146137 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7727 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40376 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112046 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7764 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6496 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112420 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28528 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5895 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117921 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61692 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7780 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36029 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7524 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71333 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7745 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7623 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/style/BorderData.h:
```diff
@@ -107,68 +108,61 @@ class BorderData {
-
-    Style::BorderImage& image() { return m_image; }
-    const Style::BorderImage& image() const { return m_image; }
+    BorderValue& left() { return edges.left(); }
+    BorderValue& right() { return edges.right(); }
+    BorderValue& top() { return edges.top(); }
+    BorderValue& bottom() { return edges.bottom(); }
```
Do all these functions that return references need `LIFETIME_BOUND` macros?

weinig replied:
Yeah, as do all the others in that file. 

Can’t wait for the new clang option to recommend missing lifetime attributes!

darinadler commented on Source/WebCore/rendering/style/BorderValue.h:
```diff
@@ -31,37 +31,24 @@
+struct BorderValue {
+    Style::Color color { Style::Color::currentColor() };
+    Style::LineWidth width { Style::LineWidth::Length { 3.0f } };
+    PREFERRED_TYPE(BorderStyle) unsigned style : 4 { static_cast<unsigned>(BorderStyle::None) };
```
Do we want a getter with the `static_cast` in cases like this?

weinig replied:
Most users will want to use RenderStyle for this which has the cast generated. This matches all the other Style*Data types. 

I have been thinking of making all the Style*Data types that have these packed enums expose getter functions with a cast, but I want to do it with a consistent naming scheme.

weinig committed:
[RenderStyleGen] Make BorderImage/MaskBorder storage more consistent with other RenderStyle types
https://bugs.webkit.org/show_bug.cgi?id=304128

Reviewed by Darin Adler.

Style::BorderImage and Style::MaskImage had a non-traditional
storage implementation by using a private DataRef value. This
required using `friend` to participate in autogeneration, but
that is not going to scale with additional generation. Instead,
we extract the internal Data types into new `StyleBorderImageData`
and `StyleMaskImageData` types, and invert the relationship.

This also allows us to get rid of the custom style building
code for border image and mask image, and will make it easier
for future automation efforts.

While doing this, I also made the members of the border types public
(and remove the m_ prefixes). We want to treat RenderStyle data
types as pure storage going forward, so being public makes sense.

Did a little cleanup RenderStyle reset functions to remove unused
ones and make the rest call to a set function.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/layout/formattingContexts/block/BlockMarginCollapse.cpp:
* Source/WebCore/page/LocalFrameView.cpp:
* Source/WebCore/rendering/OutlinePainter.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderTable.cpp:
* Source/WebCore/rendering/RenderTableSection.cpp:
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/RenderTreeAsText.cpp:
* Source/WebCore/rendering/style/BorderData.cpp:
* Source/WebCore/rendering/style/BorderData.h:
* Source/WebCore/rendering/style/BorderValue.cpp:
* Source/WebCore/rendering/style/BorderValue.h:
* Source/WebCore/rendering/style/CollapsedBorderValue.h:
* Source/WebCore/rendering/style/OutlineValue.cpp:
* Source/WebCore/rendering/style/OutlineValue.h:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleBase.h:
* Source/WebCore/rendering/style/RenderStyleBaseInlines.h:
* Source/WebCore/rendering/style/RenderStyleBaseSetters.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStylePropertiesGettersCustom.cpp:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleBackgroundData.cpp:
* Source/WebCore/rendering/style/StyleBorderImageData.cpp: Added.
* Source/WebCore/rendering/style/StyleBorderImageData.h: Copied from Source/WebCore/rendering/style/RenderStylePropertiesGettersCustom.cpp.
* Source/WebCore/rendering/style/StyleMaskBorderData.cpp: Added.
* Source/WebCore/rendering/style/StyleMaskBorderData.h: Copied from Source/WebCore/rendering/style/RenderStylePropertiesGettersCustom.cpp.
* Source/WebCore/rendering/style/StyleMultiColData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleSurroundData.cpp:
* Source/WebCore/style/StyleBuilder.cpp:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleChangedAnimatableProperties.cpp:
* Source/WebCore/style/StyleDifference.cpp:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/values/backgrounds/StyleBorderImage.cpp:
* Source/WebCore/style/values/backgrounds/StyleBorderImage.h:
* Source/WebCore/style/values/backgrounds/StyleBorderImageWidth.h:
* Source/WebCore/style/values/masking/StyleMaskBorder.cpp:
* Source/WebCore/style/values/masking/StyleMaskBorder.h:
* Source/WebCore/style/values/non-standard/StyleWebKitBoxReflect.cpp:

Canonical link: https://commits.webkit.org/304482@main
